crates/runtime/src/component/resources.rs

```diff
@@ -294,36 +298,42 @@ impl ResourceTable {
             Slot::Free { next } => next,
             _ => unreachable!(),
         };
-        u32::try_from(ret).unwrap()
+
+        // The component model reserves index 0 as never allocatable so add one
+        // to the table index to start the numbering at 1 instead. Also note
+        // that the component model places an upper-limit per-table on the
+        // maximum allowed index.
+        let ret = ret + 1;
+        if ret >= 1 << 30 {
+            bail!("cannot allocate another handle: index overflow");
+        }
+        Ok(ret)
     }
 
     fn rep(&self, idx: u32) -> Result<u32> {
-        match usize::try_from(idx).ok().and_then(|i| self.slots.get(i)) {
+        // NB: `idx` is decremented by one to account for the `+1` above during
+        // allocation.
+        let table_index = idx.checked_sub(1).and_then(|i| usize::try_from(i).ok());
+        match table_index.and_then(|i| self.slots.get(i)) {
```

Comment:
The fetching of a slot is finicky enough that I wonder if it makes sense to move it to its own function so if there is ever another need for that logic the chances of bad copy/paste can be minimized.